src/ucs/async/async_fwd.h

```diff
@@ -1,12 +1,119 @@
+* Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
 *
 * See file LICENSE for terms.
 */
 
 #ifndef UCS_ASYNC_FWD_H
 #define UCS_ASYNC_FWD_H
 
+#include <ucs/config/types.h>
+#include <ucs/time/time_def.h>
+
+
 typedef struct ucs_async_context ucs_async_context_t;
 
+
+/**
+ * @ingroup UCS_RESOURCE
+ *
+ * Async event callback.
+ *
+ * @param id           Event id (timer or file descriptor).
+ * @param arg          User-defined argument.
+ */
+typedef void (*ucs_async_event_cb_t)(int id, void *arg);
+
+
+/**
+ * @ingroup UCS_RESOURCE
+ *
+ * Register a file descriptor for monitoring (call handler upon events).
+ * Every fd can have only one handler.
+ *
+ * @param mode            Thread or signal.
+ * @param event_fd        File descriptor to set handler for.
+ * @param events          Events to wait on (POLLxx/EPOLLxx bits).
+ * @param cb              Callback function to execute.
+ * @param arg             Argument to callback.
+ * @param async           Async context to which events are delivered.
+ *                        If NULL, safety is up to the user.
+ */
+ucs_status_t ucs_async_set_event_handler(ucs_async_mode_t mode, int event_fd,
+                                         int events, ucs_async_event_cb_t cb,
+                                         void *arg, ucs_async_context_t *async);
+
+
+/**
+ * @ingroup UCS_RESOURCE
+ *
+ * Add timer handler.
+ *
+ * @param mode            Thread or signal.
+ * @param interval        Timer interval.
+ * @param cb              Callback function to execute.
+ * @param arg             Argument to callback.
+ * @param async           Async context to which events are delivered.
+ *                        If NULL, safety is up to the user.
+ * @param timer_id_p      Filled with timer id.
+ */
+ucs_status_t ucs_async_add_timer(ucs_async_mode_t mode, ucs_time_t interval,
+                                 ucs_async_event_cb_t cb, void *arg,
+                                 ucs_async_context_t *async, int *timer_id_p);
+
+
+/**
+ * @ingroup UCS_RESOURCE
+ *
+ * Remove an event handler (Timer or event file).
+ *
+ * @param id        Timer/FD to remove.
+ * @param sync      If nonzero, wait until the handler for this event is not
+ *                  running anymore. Cannot be used in the context of the event
+ *                  handler itself because it would deadlock.
+ */
+ucs_status_t ucs_async_remove_handler(int id, int sync);
+
+
+/**
+ * @ingroup UCS_RESOURCE
+ * @brief Allocate and initialize an asynchronous execution context
+ *
+ * Allocate and initialize an asynchronous execution context.
+ * This can be used to ensure safe event delivery.
+ *
+ * @param async           Event context pointer to initialize.
+ * @param mode            Either to use signals or epoll threads to wait.
+ *
+ * @return Error code as defined by @ref ucs_status_t.
+ */
+ucs_status_t ucs_async_context_init(ucs_async_context_t **async,
```

Comment:
1. sure, i will change it if we find a conclusion for >2<
2. I think it broke the UCS public interface. I kept function `ucs_async_context_init` as is for public and create `ucs_async_context_init_internal` for leave any public interface untouchable. For example, in `uct_hello_world` we used `ucs_async_context_init`. I assume, unknown number of external users have this function in their programs. If we replace `ucs_async_context_init` by `ucp_async_create` they have to change their programs.
In other hand, using `ucs_async_context_init_internal` isn't comfortable for UCX internally.
Do you agree to change UCS public interface?

Also, I would like to ask @shamisp opinion about this. The main goal for all these changes is to move `ucs_async_context_t` (and many related types) from user level to internal and leave this type as opaque for the user. The problem - we have to allocate memory for this type since the user have no information about the type length.

From my point of view, changing UCS public interface, is reasonable for this particular case.